### PR TITLE
Refactor resending order confirmation email and cover it in API

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | no                                                       |
| New feature?    | a bit                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | yes  |
| License         | MIT                                                          |

### CODE REVIEW

File: src/Sylius/Bundle/AdminBundle/Action/ResendOrderConfirmationEmailAction.php [hunk 1]
```diff
@@ -31,7 +32,7 @@ final class ResendOrderConfirmationEmailAction
 {
     public function __construct(
         private OrderRepositoryInterface $orderRepository,
-        private OrderEmailManagerInterface $orderEmailManager,
+        private OrderEmailManagerInterface|ResendOrderConfirmationEmailDispatcherInterface $orderEmailManagerOrResendOrderConfirmationDispatcher,
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/AdminBundle/Action/ResendOrderConfirmationEmailAction.php]: Due to named arguments mechanism, we shouldn't touch the parameter's name. Also, we should inform it'll be changed in the future.

**Review by jakubtobiasz — DISMISSED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 2]
```diff
@@ -67,6 +67,21 @@
                     <attribute name="groups">shop:order:read</attribute>
                 </attribute>
             </collectionOperation>
+
+            <collectionOperation name="admin_resend_order_confirmation">
+                <attribute name="method">POST</attribute>
+                <attribute name="path">/admin/orders/resend-order-confirmation-email</attribute>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml]: Why don't you use the order token as a parameter?
```suggestion
                <attribute name="path">/admin/orders/{tokenValue}/resend-confirmation-email</attribute>
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 3]
```diff
@@ -139,6 +142,22 @@ public function iSpecifyFilterDateToAs(string $dateTime): void
...
+    {
+        /** @var RequestInterface $request */
+        $request = $this->requestFactory->custom(
+            \sprintf('%s/admin/orders/resend-order-confirmation-email', $this->apiUrlPrefix),
+            HttpRequest::METHOD_POST,
+        );
+
+        $request->setContent(['orderToken' => $this->sharedStorage->get('order')->getTokenValue()]);
+
+        $this->client->executeCustomRequest($request);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: You should be able to use here `$this->client->customItemAction(...)` and not have to add an API prefix to the context

File: src/Sylius/Bundle/CoreBundle/MessageHandler/Admin/ResendOrderConfirmationEmailHandler.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    public function __construct(
+        private OrderEmailManagerInterface $orderEmailManager,
+        private RepositoryInterface $orderRepository,
+    ) {
+    }
+
+    public function __invoke(ResendOrderConfirmationEmail $resendOrderConfirmation): void
+    {
+        /** @var OrderInterface|null $order */
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $resendOrderConfirmation->orderToken()]);
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/MessageHandler/Admin/ResendOrderConfirmationEmailHandler.php]: Shouldn't we also limit the states here?
It'd be weird to allow resending confirmation for fulfilled or cancelled orders

File: src/Sylius/Bundle/CoreBundle/EmailManager/OrderEmailManager.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+        $this->emailSender->send(
+            Emails::ORDER_CONFIRMATION_RESENT,
+            [$email],
+            [
+                'order' => $order,
+                'channel' => $order->getChannel(),
+                'localeCode' => $order->getLocaleCode(),
+            ],
+            [],
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/EmailManager/OrderEmailManager.php]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 6]
```diff
@@ -67,6 +67,21 @@
                     <attribute name="groups">shop:order:read</attribute>
                 </attribute>
             </collectionOperation>
+
+            <collectionOperation name="admin_resend_order_confirmation">
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml]: Why is this a collection operation?

File: features/order/managing_orders/resending_order_confirmation_email.feature [hunk 7]
```diff
@@ -16,14 +16,14 @@ Feature: Resending an order confirmation email for a chosen order
...
+    @ui @email @api
     Scenario: Resending a confirmation email for a given order
         When I view the summary of the order "#00000666"
         And I resend the order confirmation email
         Then I should be notified that the order confirmation email has been successfully resent to the customer
         And an email with the confirmation of the order "#00000666" should be sent to "lucy@teamlucifer.com"
 
-    @ui @email
+    @ui @email @api
     Scenario: Sending a confirmation email after shipping an order in different locale than the default one
```
Comment (NoResponseMate) [inline features/order/managing_orders/resending_order_confirmation_email.feature:27]: This scenario sounds off

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 8]
```diff
@@ -312,6 +331,14 @@ public function iShouldSeeASingleOrderInTheList(int $number = 1): void
         Assert::same($this->responseChecker->countCollectionItems($this->client->getLastResponse()), $number);
     }
 
+    /**
+     * @Then /^I should be notified that the (order|shipment) confirmation email has been successfully resent to the customer$/
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
     * @Then /^I should be notified that the order confirmation email has been successfully resent to the customer$/
```
Context leak between PRs, you'll most likely have conflicts either way

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml (hunk 6 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml]:
```suggestion
            <collectionOperation name="admin_resend_confirmation">
```
as this is operation for Order resource

File: src/Sylius/Bundle/AdminBundle/Action/ResendOrderConfirmationEmailAction.php [hunk 9]
```diff
@@ -45,6 +46,17 @@ public function __construct(
                 RequestStack::class,
             );
         }
+
+        if ($this->orderEmailManager instanceof OrderEmailManagerInterface) {
+            trigger_deprecation(
+                'sylius/admin-bundle',
+                '1.13',
+                'Passing an instance of %s as constructor argument for %s is deprecated and will be removed in Sylius 2.0. Pass an instance of %s instead.',
```
Comment (Wojdylak) [inline src/Sylius/Bundle/AdminBundle/Action/ResendOrderConfirmationEmailAction.php:54]: I noticed that you suggested changing the name of the parameter. Maybe we should add information here stating that in version 2.0, the name of `orderEmailManager` will be changed to `resendOrderConfirmationEmailDispatcher`. WDYT?

**Review by Wojdylak — DISMISSED** [review]

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 10]
```diff
@@ -51,6 +52,7 @@ public function __construct(
         private SharedStorageInterface $sharedStorage,
         private SharedSecurityServiceInterface $sharedSecurityService,
         private SectionAwareIriConverterInterface $sectionAwareIriConverter,
+        private string $apiUrlPrefix,
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: It's not needed anymore if I see correcty:
```suggestion
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 11]
```diff
@@ -312,6 +327,14 @@ public function iShouldSeeASingleOrderInTheList(int $number = 1): void
         Assert::same($this->responseChecker->countCollectionItems($this->client->getLastResponse()), $number);
     }
 
+    /**
+     * @Then /^I should be notified that the order confirmation email has been successfully resent to the customer$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
     * @Then I should be notified that the order confirmation email has been successfully resent to the customer
```

File: src/Sylius/Behat/Resources/config/services/contexts/api/admin.xml [hunk 12]
```diff
@@ -228,6 +228,7 @@
             <argument type="service" id="sylius.behat.shared_storage" />
             <argument type="service" id="sylius.behat.api.shared_security" />
             <argument type="service" id="sylius.behat.section_iri_converter" />
+            <argument>%sylius.security.new_api_route%</argument>
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/services/contexts/api/admin.xml]:
```suggestion
```

File: src/Sylius/Bundle/CoreBundle/Message/Admin/ResendOrderConfirmationEmail.php [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Message\Admin;
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Message/Admin/ResendOrderConfirmationEmail.php]: According to the discussion on the similar [PR](https://github.com/Sylius/Sylius/pull/15715#discussion_r1447817352), I would be for:
```suggestion
namespace Sylius\Bundle\CoreBundle\Message;
```

File: src/Sylius/Bundle/CoreBundle/MessageHandler/Admin/ResendOrderConfirmationEmailHandler.php [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\MessageHandler\Admin;
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/MessageHandler/Admin/ResendOrderConfirmationEmailHandler.php]: As above:
```suggestion
namespace Sylius\Bundle\CoreBundle\MessageHandler;
```

File: src/Sylius/Bundle/CoreBundle/Validator/Constraints/ResendsOrderConfirmationEmailWithValidOrderState.php [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+final class ResendsOrderConfirmationEmailWithValidOrderState extends Constraint
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Validator/Constraints/ResendsOrderConfirmationEmailWithValidOrderState.php]:
```suggestion
final class ResendOrderConfirmationEmailWithValidOrderState extends Constraint
```
or even
```suggestion
final class OrderConfirmationEmailWithValidOrderState extends Constraint
```

File: src/Sylius/Bundle/CoreBundle/Message/Admin/ResendOrderConfirmationEmail.php (hunk 13 above)
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Message/Admin/ResendOrderConfirmationEmail.php]: Same question as in parallel PR https://github.com/Sylius/Sylius/pull/15715#discussion_r1447817352

CoreBundle doesn't depend or know about any admins, so `Admin` cannot be part of namespace.

File: src/Sylius/Bundle/CoreBundle/MessageDispatcher/ResendOrderConfirmationEmailDispatcher.php [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+use Sylius\Bundle\CoreBundle\Message\Admin\ResendOrderConfirmationEmail;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+final class ResendOrderConfirmationEmailDispatcher implements ResendOrderConfirmationEmailDispatcherInterface
+{
+    public function __construct(private MessageBusInterface $messageBus)
+    {
+    }
+
+    public function dispatch(string $orderToken): void
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/MessageDispatcher/ResendOrderConfirmationEmailDispatcher.php]: Same as in parallel order shipment confirmation PR
https://github.com/Sylius/Sylius/pull/15715#discussion_r1447809855
https://github.com/Sylius/Sylius/pull/15715#discussion_r1448837020

File: src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_resendEmail.html.twig [hunk 17]
```diff
@@ -1,6 +1,8 @@
-<div class="ui segment">
-    {% set path = path('sylius_admin_order_resend_confirmation_email', {'id': order.id, '_csrf_token': csrf_token(order.id)}) %}
-    <a href="{{ path }}" class="ui icon labeled tiny fluid button" {{ sylius_test_html_attribute('resend-order-confirmation-email') }}>
-        <i class="send icon"></i> {{ 'sylius.ui.resend_the_order_confirmation_email'|trans }}
-    </a>
-</div>
+{% if order.state == 'new'%}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_resendEmail.html.twig]: I'm not sure if this is the only state in which we should resent the email, what about fulfilled? 🤔

Comment (TheMilek) [reply src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_resendEmail.html.twig]: Fulfilled state is when we also ship the order, so in this case I would be after sticking only to the new state 🤔

File: src/Sylius/Bundle/CoreBundle/MessageHandler/ResendOrderConfirmationEmailHandler.php [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+        private RepositoryInterface $orderRepository,
+    ) {
+    }
+
+    public function __invoke(ResendOrderConfirmationEmail $resendOrderConfirmation): void
+    {
+        /** @var OrderInterface|null $order */
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $resendOrderConfirmation->getOrderTokenValue()]);
+        if ($order === null) {
+            throw new NotFoundHttpException(sprintf('The order with id %s has not been found', $resendOrderConfirmation->getOrderTokenValue()));
```
Comment (Wojdylak) [inline src/Sylius/Bundle/CoreBundle/MessageHandler/ResendOrderConfirmationEmailHandler.php]:
```suggestion
            throw new NotFoundHttpException(sprintf('The order with tokenValue %s has not been found', $resendOrderConfirmation->getOrderTokenValue()));
```

**Review by Wojdylak — DISMISSED** [review]

File: src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_resendEmail.html.twig (hunk 17 above)
Comment (TheMilek) [reply src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_resendEmail.html.twig]: I moved it to separate PR https://github.com/Sylius/Sylius/pull/15735

**Review by Wojdylak — APPROVED** [review]

File: src/Sylius/Bundle/AdminBundle/Action/ResendOrderConfirmationEmailAction.php [hunk 19]
```diff
@@ -45,6 +46,24 @@ public function __construct(
...
+                OrderEmailManagerInterface::class,
+                self::class,
+                ResendOrderConfirmationEmailDispatcherInterface::class,
+            );
+
+            trigger_deprecation(
+                'sylius/admin-bundle',
+                '1.13',
+                'The argument name of %s is deprecated and will be renamed to $resendOrderConfirmationEmailDispatcher in Sylius 2.0.',
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/Action/ResendOrderConfirmationEmailAction.php:63]:
```suggestion
                'The argument name of %s is deprecated and will be renamed from X to $resendOrderConfirmationEmailDispatcher in Sylius 2.0.',
```

**Review by NoResponseMate — APPROVED** [review]

**Comment by NoResponseMate** [comment]
Thank you, @TheMilek!

**Pull request merged**